### PR TITLE
[Scale] - Remove d3 scale input to constructor

### PR DESCRIPTION
**Pull request by bluong**
We shouldn't be exposing the fact that D3 backs our scales in our API.  That's purely an implementation detail

### CODE REVIEW

File: src/scales/categoryScale.ts [hunk 1]
```diff
@@ -3,7 +3,6 @@
 module Plottable {
 export module Scales {
   export class Category extends Scale<string, number> {
-    protected _d3Scale: D3.Scale.OrdinalScale;
```
Comment (aicioara) [inline src/scales/categoryScale.ts]: Is there a good reason why this was deleted? I see two reasons to keep it:
- When reading code, you don't have to navigate to a different file to find the variable
- You don't need those ugly casts (which get uglier in the next files) every time you want to use the scale.

If we decide to keep it, this comment applies to all future files (including those where the `protected _d3Scale` was not present in the first place (for consistency)

Comment (bluong) [inline src/scales/categoryScale.ts]: So, unless I'm mistaken the technique that is used here is called "shadowing" where `this._d3Scale` in this context shares the same reference as the one in the parent class.

One problem I have with this technique is that if the parent class were to change it's instance variable name, we can easily enter a scenario where the child class's shadowed instance variable is accidentally not also renamed and thus we have 2 instance variables when we should have one.

To give an example, I think this has caused a bug in the past that was really hard to detect.  The `_isVertical` flag I believe used to be named `isVertical`.  However, for some reason it was used in quite a few places in the hierarchy.  Someone (probably I) changed `isVertical` to `_isVertical` and forgot to change this fact on some of the child classes.  Thankfully I think we wrote our code such that we didn't run into user-facing bugs, but this sounds like something that could easily be dangerous in a quiet way which is quite scary to me.

What do you think?

Comment (aicioara) [inline src/scales/categoryScale.ts]: That's why Java has `@Override`. Hmm...

Comment (aicioara) [inline src/scales/categoryScale.ts]: Can we go around this? I really don't like the casts. Do something like
`this._d3Scale = <cast_here>super._d3Scale`
This will solve that issue

File: src/components/plots/gridPlot.ts [hunk 2]
```diff
@@ -59,9 +59,10 @@ export module Plots {
         super.x(<number | _Accessor> x);
       } else {
         super.x(<any | _Accessor> x, scale);
+        var catScale = (<Scales.Category> <any> scale);
```
Comment (aicioara) [inline src/components/plots/gridPlot.ts]: move it inside the `if`

File: src/components/plots/gridPlot.ts [hunk 3]
```diff
@@ -80,9 +81,10 @@ export module Plots {
         super.y(<number | _Accessor> y);
       } else {
         super.y(<any | _Accessor> y, scale);
+        var catScale = (<Scales.Category> <any> scale);
```
Comment (aicioara) [inline src/components/plots/gridPlot.ts]: same

File: src/components/plots/gridPlot.ts [hunk 4]
```diff
@@ -22,10 +22,10 @@ export module Plots {
 
       // The x and y scales should render in bands with no padding for category scales
       if (xScale instanceof Scales.Category) {
-        xScale.innerPadding(0).outerPadding(0);
+        <Scales.Category> xScale.innerPadding(0).outerPadding(0);
```
Comment (aicioara) [inline src/components/plots/gridPlot.ts]: Why was this cast not necessary before?

File: src/scales/interpolatedColorScale.ts [hunk 5]
```diff
@@ -10,7 +10,7 @@ export module Scales {
    *
    * By default it generates a linear scale internally.
    */
-  export class InterpolatedColor extends Scale<number, string> {
+  export class InterpolatedColor extends Scale<number, number> {
```
Comment (aicioara) [inline src/scales/interpolatedColorScale.ts]: Why this change?

File: src/scales/quantitativeScale.ts [hunk 6]
```diff
@@ -33,29 +18,23 @@ module Plottable {
...
      */
     public invert(value: number): D {
-      return <any> this._d3Scale.invert(value);
-    }
-
-    public domain(): D[];
-    public domain(values: D[]): QuantitativeScale<D>;
-    public domain(values?: D[]): any {
-      return super.domain(values); // need to override type sig to enable method chaining:/
+      return null;
```
Comment (aicioara) [inline src/scales/quantitativeScale.ts]: Isn't `throw` a better alternative?

File: src/scales/scale.ts [hunk 7]
```diff
@@ -93,7 +89,7 @@ module Plottable {
      * @returns {R} The range value corresponding to the supplied domain value.
      */
     public scale(value: D): R {
-      return this._d3Scale(value);
+      return null;
```
Comment (aicioara) [inline src/scales/scale.ts]: Same, isn't throw a better alternative?

File: src/scales/scale.ts [hunk 8]
```diff
@@ -122,19 +118,23 @@ module Plottable {
...
         this._domainModificationInProgress = true;
-        this._d3Scale.domain(values);
+        this._setActualDomain(values);
         this._dispatchUpdate();
         this._domainModificationInProgress = false;
       }
     }
 
+    protected _setActualDomain(values: D[]) {
+      // Subclasses should override
```
Comment (aicioara) [inline src/scales/scale.ts]: Throw?

File: src/scales/scale.ts [hunk 9]
```diff
@@ -158,13 +158,21 @@ module Plottable {
...
         return this;
       }
     }
 
+    protected _getRange(): R[] {
+      return [];
+    }
+
+    protected _setRange(values: R[]) {
+      // Subclasses should override
```
Comment (aicioara) [inline src/scales/scale.ts]: Throw?

File: src/scales/timeScale.ts [hunk 10]
```diff
@@ -47,6 +46,38 @@ export module Scales {
...
+    public invert(value: number) {
+      return this._d3Scale.invert(value);
+    }
+
+    public getDefaultTicks(): Date[] {
+      return this._d3Scale.ticks(QuantitativeScale._DEFAULT_NUM_TICKS);
+    }
+
+    public _niceDomain(domain: Date[], count?: number) {
+      return Utils.D3Scale.niceDomain(<D3.Scale.QuantitativeScale> <D3.Scale.Scale> this._d3Scale, domain, count);
```
Comment (aicioara) [inline src/scales/timeScale.ts]: Is there any workaround for this? Double casting is so ugly.

File: test/components/numericAxisTests.ts [hunk 11]
```diff
@@ -307,8 +307,8 @@ describe("NumericAxis", () => {
     var yLabel = new Plottable.Components.Label("LABEL", "left");
     yLabel.classed(Plottable.Components.Label.AXIS_LABEL_CLASS, true);
     var barPlot = new Plottable.Plots.Bar(xScale, yScale);
-    barPlot.x((d) => d.x, xScale);
-    barPlot.y((d) => d.y, yScale);
+    barPlot.x((d: any) => d.x, xScale);
+    barPlot.y((d: any) => d.y, yScale);
```
Comment (aicioara) [inline test/components/numericAxisTests.ts]: Why is this needed now and not before?

**Comment by aicioara** [comment]
Regarding the `throw`s, at least log a warning, if not really throw. Returning `null` is ugly
Regarding the `why is it needed`, I was just curious

File: src/scales/timeScale.ts (hunk 10 above)
Comment (bluong) [inline src/scales/timeScale.ts]: Unless I'm mistaken `D3.Scale.TimeScale` is actually a child of `D3.Scale.QuantitativeScale` but the d3.d.ts file does not say this point.  Let me check/verify

File: src/components/plots/gridPlot.ts (hunk 4 above)
Comment (bluong) [inline src/components/plots/gridPlot.ts]: Unsure...

File: test/components/numericAxisTests.ts (hunk 11 above)
Comment (bluong) [inline test/components/numericAxisTests.ts]: Compilation error now.  I'm actually unsure why it fails now and not before

File: src/scales/interpolatedColorScale.ts (hunk 5 above)
Comment (bluong) [inline src/scales/interpolatedColorScale.ts]: After diving into the code, it forced me to make this contract.  Looking at it again, this does look odd.  @rcchen do you have any context on why the compiler would force me to make this decision?

File: src/scales/scale.ts [hunk 12]
```diff
@@ -122,19 +118,23 @@ module Plottable {
...
       if (!this._domainModificationInProgress) {
         this._domainModificationInProgress = true;
-        this._d3Scale.domain(values);
+        this._setActualDomain(values);
         this._dispatchUpdate();
         this._domainModificationInProgress = false;
       }
     }
 
+    protected _setActualDomain(values: D[]) {
```
Comment (jtlan) [inline src/scales/scale.ts]: What's this? This name is really confusing, since there's already a `_setDomain()` method.

Comment (bluong) [inline src/scales/scale.ts]: This is specifically to set the domain of usually the backing D3 scale.  I was completely unsure of the name as it's basically `_setDomain()` of the backing scale but I wasn't sure of how to go about it.  I mainly didn't want to rewrite the previous implementation of `setDomain()` over and over as that can lead to inconsistencies.

File: src/scales/categoryScale.ts [hunk 13]
```diff
@@ -142,7 +142,23 @@ export module Scales {
...
+    protected _setBackingDomain(values: string[]) {
+      this._d3Scale.domain(values);
+    }
+
+    protected _getRange() {
+      return this._d3Scale.range();
+    }
+
+    protected _setRange(values: number[]) {
+      this._d3Scale.range(values);
```
Comment (aicioara) [inline src/scales/categoryScale.ts]: This looks so Java-ish (i.e. verbose)... but sure

File: src/scales/linearScale.ts [hunk 14]
```diff
@@ -13,15 +14,46 @@ export module Scales {
...
+      return this._d3Scale.invert(value);
+    }
+
+    public getDefaultTicks(): number[] {
+      return this._d3Scale.ticks(QuantitativeScale._DEFAULT_NUM_TICKS);
+    }
+
+    public _niceDomain(domain: number[], count?: number): number[] {
+      return Utils.D3Scale.niceDomain(this._d3Scale, domain, count);
+    }
```
Comment (aicioara) [inline src/scales/linearScale.ts]: Move `public` above `protected`

File: src/scales/linearScale.ts [hunk 15]
```diff
@@ -13,15 +14,46 @@ export module Scales {
...
+    public invert(value: number) {
+      return this._d3Scale.invert(value);
+    }
+
+    public getDefaultTicks(): number[] {
+      return this._d3Scale.ticks(QuantitativeScale._DEFAULT_NUM_TICKS);
+    }
+
+    public _niceDomain(domain: number[], count?: number): number[] {
+      return Utils.D3Scale.niceDomain(this._d3Scale, domain, count);
```
Comment (aicioara) [inline src/scales/linearScale.ts]: `public` underscored?

File: src/scales/modifiedLogScale.ts [hunk 16]
```diff
@@ -97,6 +99,10 @@ export module Scales {
       super._setDomain(transformedDomain);
     }
 
+    protected _setBackingDomain(values: number[]) {
```
Comment (aicioara) [inline src/scales/modifiedLogScale.ts]: Move protected after public

File: src/scales/modifiedLogScale.ts [hunk 17]
```diff
@@ -204,6 +210,18 @@ export module Scales {
...
+      return this._d3Scale.range();
+    }
+
+    protected _setRange(values: number[]) {
+      this._d3Scale.range(values);
+    }
+
+    public getDefaultTicks(): number[] {
+      return this._d3Scale.ticks(QuantitativeScale._DEFAULT_NUM_TICKS);
+    }
```
Comment (aicioara) [inline src/scales/modifiedLogScale.ts]: `public` < `protected`

File: src/scales/quantitativeScale.ts [hunk 18]
```diff
@@ -33,29 +18,23 @@ module Plottable {
...
      */
     public invert(value: number): D {
-      return <any> this._d3Scale.invert(value);
-    }
-
-    public domain(): D[];
-    public domain(values: D[]): QuantitativeScale<D>;
-    public domain(values?: D[]): any {
-      return super.domain(values); // need to override type sig to enable method chaining:/
+      throw new Error("subclass should override _niceDomain");
```
Comment (aicioara) [inline src/scales/quantitativeScale.ts]: Capitalize Sublcass

File: src/scales/quantitativeScale.ts [hunk 19]
```diff
@@ -33,29 +18,23 @@ module Plottable {
...
+      }
+      super._setDomain(values);
     }
 
     /**
      * Gets ticks generated by the default algorithm.
      */
     public getDefaultTicks(): D[] {
-        return this._d3Scale.ticks(QuantitativeScale._DEFAULT_NUM_TICKS);
+      throw new Error("subclass should override _niceDomain");
```
Comment (aicioara) [inline src/scales/quantitativeScale.ts]: capitalize + function name

File: src/scales/quantitativeScale.ts [hunk 20]
```diff
@@ -72,7 +51,7 @@ module Plottable {
      * numbers.
      */
     public _niceDomain(domain: D[], count?: number): D[] {
-      return this._d3Scale.copy().domain(domain).nice(count).domain();
+      throw new Error("subclass should override _niceDomain");
```
Comment (aicioara) [inline src/scales/quantitativeScale.ts]: capitalize

File: src/scales/scale.ts [hunk 21]
```diff
@@ -93,7 +89,7 @@ module Plottable {
      * @returns {R} The range value corresponding to the supplied domain value.
      */
     public scale(value: D): R {
-      return this._d3Scale(value);
+      throw new Error("subclasses should override functionality");
```
Comment (aicioara) [inline src/scales/scale.ts]: capitalize. Now we're not consistent, everywhere above it was stated the function name in the error

File: src/scales/scale.ts [hunk 22]
```diff
@@ -122,19 +118,23 @@ module Plottable {
       }
     }
 
-    protected _getDomain() {
-      return this._d3Scale.domain();
+    protected _getDomain(): D[] {
```
Comment (aicioara) [inline src/scales/scale.ts]: I don't think you need the return type in the function definition. Check the `plottable.d.ts` if it is autogenerated anyway

File: src/scales/scale.ts [hunk 23]
```diff
@@ -122,19 +118,23 @@ module Plottable {
...
+      throw new Error("subclasses should override _getDomain");
```
Comment (aicioara) [inline src/scales/scale.ts]: Capitalize

File: src/scales/scale.ts [hunk 24]
```diff
@@ -122,19 +118,23 @@ module Plottable {
...
         this._domainModificationInProgress = true;
-        this._d3Scale.domain(values);
+        this._setBackingDomain(values);
         this._dispatchUpdate();
         this._domainModificationInProgress = false;
       }
     }
 
+    protected _setBackingDomain(values: D[]) {
+      throw new Error("subclasses should override _setBackingDomain");
```
Comment (aicioara) [inline src/scales/scale.ts]: Capitalize

File: src/scales/scale.ts [hunk 25]
```diff
@@ -158,13 +158,21 @@ module Plottable {
...
-        return this._d3Scale.range();
+        return this._getRange();
       } else {
-        this._d3Scale.range(values);
+        this._setRange(values);
         return this;
       }
     }
 
+    protected _getRange(): R[] {
```
Comment (aicioara) [inline src/scales/scale.ts]: Not sure if the return type is needed (as above)

File: src/scales/scale.ts [hunk 26]
```diff
@@ -158,13 +158,21 @@ module Plottable {
...
+      throw new Error("subclasses should override _getRange");
+    }
+
+    protected _setRange(values: R[]) {
```
Comment (aicioara) [inline src/scales/scale.ts]: Not sure if the return type is needed (as above)

File: src/scales/timeScale.ts [hunk 27]
```diff
@@ -47,6 +46,38 @@ export module Scales {
...
+      return this._d3Scale.invert(value);
+    }
+
+    public getDefaultTicks(): Date[] {
+      return this._d3Scale.ticks(QuantitativeScale._DEFAULT_NUM_TICKS);
+    }
+
+    public _niceDomain(domain: Date[], count?: number) {
+      return Utils.D3Scale.niceDomain(this._d3Scale, domain, count);
+    }
```
Comment (aicioara) [inline src/scales/timeScale.ts]: Same issue with `public` > `protected`, public underscored. That's why I said it is super verbose and prone to repetition

File: src/scales/timeScale.ts [hunk 28]
```diff
@@ -47,6 +46,38 @@ export module Scales {
...
+
+    public scale(value: Date): number {
+      return this._d3Scale(value);
+    }
+
+    protected _getDomain() {
+      return this._d3Scale.domain();
+    }
+
+    protected _setBackingDomain(values: Date[]) {
```
Comment (aicioara) [inline src/scales/timeScale.ts]: method name does not make sense. What is a backing domain?

File: src/scales/scale.ts [hunk 29]
```diff
@@ -122,19 +118,23 @@ module Plottable {
...
       if (!this._domainModificationInProgress) {
         this._domainModificationInProgress = true;
-        this._d3Scale.domain(values);
+        this._setBackingDomain(values);
         this._dispatchUpdate();
         this._domainModificationInProgress = false;
       }
     }
 
+    protected _setBackingDomain(values: D[]) {
```
Comment (aicioara) [inline src/scales/scale.ts]: All of these should be documented in the superclass (here). As a developer I am not even sure what is the contract I am getting into if I decide to extend the classes and I have to implement these methods.

File: src/scales/linearScale.ts (hunk 14 above)
Comment (bluong) [inline src/scales/linearScale.ts]: We currently don't have a rule surrounding this.  I'm also unsure about the usefulness of this rule...

File: src/scales/linearScale.ts (hunk 15 above)
Comment (bluong) [inline src/scales/linearScale.ts]: `_niceDomain` has not been cleared from the codebase.  I believe @jtlan 's `Domainer` changes is planning to address this

File: src/scales/modifiedLogScale.ts (hunk 16 above)
Comment (bluong) [inline src/scales/modifiedLogScale.ts]: As above

File: src/scales/modifiedLogScale.ts (hunk 17 above)
Comment (bluong) [inline src/scales/modifiedLogScale.ts]: As above

File: src/scales/timeScale.ts (hunk 27 above)
Comment (bluong) [inline src/scales/timeScale.ts]: As above

File: src/scales/timeScale.ts (hunk 28 above)
Comment (bluong) [inline src/scales/timeScale.ts]: Since there's gotta be something backing the domain of the scale but it is never explicitly set anywhere, the domain had to actually be set somehwere.  There is an implementation of `_setDomain()` that has to tap into how to actually set the domain of the scale, and I'm unsure of how this name should be.

Comment (aicioara) [inline src/scales/timeScale.ts]: @bluong  What is the difference between `_setDomain()` and `_setBackingDomain()` and why do we need both?

**Comment by aicioara** [comment]
Document at least `setBackingDomain()`

File: src/scales/scale.ts (hunk 29 above)
Comment (bluong) [inline src/scales/scale.ts]: I'll get to documentation after we get our API into a good state.  I'm mainly trying to focus on functionality and once we have the refactor into a somewhat stable state, I'll go document this.

File: src/scales/timeScale.ts (hunk 28 above)
Comment (bluong) [inline src/scales/timeScale.ts]: `_setDomain()` has the job of making sure that the domain of the scale is set to the according value.  However, this does not necessarily concern setting the domain of whatever backing scale.

I do note that the naming is confusing, and I'm completely open to suggestions on how to change the design of how to implement this idea.

File: src/scales/linearScale.ts [hunk 30]
```diff
@@ -13,15 +14,46 @@ export module Scales {
...
+
+    public scale(value: number) {
+      return this._d3Scale(value);
+    }
+
+    protected _getDomain() {
+      return this._d3Scale.domain();
+    }
+
+    protected _setBackingDomain(values: number[]) {
```
Comment (aicioara) [inline src/scales/linearScale.ts]: Rename to `_setBackingScaleDomain()`
